### PR TITLE
Drawing Tool

### PR DESCRIPTION
**Pull request by dangond-ptc**
Finished and polished initial pass on the spatial drawing tool.

### CODE REVIEW

File: tools/spatialDraw/DrawingManager.js [hunk 1]
```diff
@@ -0,0 +1,758 @@
...
+        });
+        const intersects = this.raycaster.intersectObjects(serializedDrawings);
+        intersects.forEach(intersect => {
+            if (intersect.object.serialized) {
+                const undoEvent = {
+                    type: 'draw',
+                    data: intersect.object.serialized
+                };
+                intersect.object.parent.remove(intersect.object);
+                this.pushEvent(undoEvent);
```
Comment (hobinjk-ptc) [inline tools/spatialDraw/DrawingManager.js:185]: Nice architectural connection!

File: tools/spatialDraw/DrawingManager.js [hunk 2]
```diff
@@ -0,0 +1,758 @@
...
+     */
+    pushEvent(event) {
+        this.eventStack.push(event);
+        this.shareUpdates();
+    }
+
+    /**
+     * Performs the most recent event in the event stack.
+     */
+    undoEvent() {
```
Comment (hobinjk-ptc) [inline tools/spatialDraw/DrawingManager.js:210]: This name doesn't really make sense to me if it's doing the latest non-erase event. Is this more like onUndoEvent/afterUndoEvent?

File: tools/spatialDraw/DrawingManager.js [hunk 3]
```diff
@@ -0,0 +1,758 @@
...
+        const position = {
+            x: (pointerEvent.pageX / window.innerWidth) * 2 - 1,
+            y: - (pointerEvent.pageY / window.innerHeight) * 2 + 1,
+        };
+
+        this.raycaster.setFromCamera(position, camera);
+
+        const ray = this.raycaster.ray;
+
+        this.position = ray.origin.clone().add(ray.direction.clone().multiplyScalar(this.offset)).applyMatrix4(camera.matrixWorld).applyMatrix4(scene.matrixWorld.clone().invert());
```
Comment (hobinjk-ptc) [inline tools/spatialDraw/DrawingManager.js:582]: This line would be a bit less daunting if it were wrapped or split up in some way

File: tools/spatialDraw/DrawingManager.js [hunk 4]
```diff
@@ -0,0 +1,758 @@
...
+        const position = {
+            x: (pointerEvent.pageX / window.innerWidth) * 2 - 1,
+            y: - (pointerEvent.pageY / window.innerHeight) * 2 + 1,
+        };
+
+        this.raycaster.setFromCamera(position, camera);
+
+        const ray = this.raycaster.ray;
+
+        this.position = ray.origin.clone().add(ray.direction.clone().multiplyScalar(offset)).applyMatrix4(camera.matrixWorld).applyMatrix4(scene.matrixWorld.clone().invert());
```
Comment (hobinjk-ptc) [inline tools/spatialDraw/DrawingManager.js:624]: Same here

File: tools/spatialDraw/DrawingManager.js [hunk 5]
```diff
@@ -0,0 +1,758 @@
...
+        }
+
+        const position = {
+            x: (pointerEvent.pageX / window.innerWidth) * 2 - 1,
+            y: - (pointerEvent.pageY / window.innerHeight) * 2 + 1,
+        };
+        this.raycaster.setFromCamera(position, camera);
+        const ray = this.raycaster.ray;
+
+        const lastOffsetPosition = ray.origin.clone().add(ray.direction.clone().multiplyScalar(this.lastOffset - this.bumpTowardsCamera)).applyMatrix4(camera.matrixWorld).applyMatrix4(scene.matrixWorld.clone().invert());
```
Comment (hobinjk-ptc) [inline tools/spatialDraw/DrawingManager.js:698]: nit: wrap

**Review by hobinjk-ptc — APPROVED** [review]
Very well-documented and structured! Some minor notes inline

File: tools/spatialDraw/index.js [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+        appActive = true;
+        scene.visible = true;
+    }
+});
+envelope.onClose(() => {
+    drawingManager.disableInteractions();
+    appActive = false;
+    scene.visible = false;
+    location.reload();
+});
```
Comment (benptc) [inline tools/spatialDraw/index.js:108]: Nice use of the envelope!

File: tools/spatialDraw/index.js [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+function resetScroll() {
+    if (window.scrollX !== 0 || window.scrollY !== 0) {
+        window.scrollTo(0, 0); // don't let keyboard events scroll the window
+    }
+    parent.postMessage(JSON.stringify({resetScroll: true}), '*');
+}
+resetScroll();
+
+function initDrawingApp() {
+    drawingManager = new DrawingManager(mainContainerObj, camera);
```
Comment (benptc) [inline tools/spatialDraw/index.js:126]: I like the way you structured this with the DrawingManager! Nice work managing the potential complexity of this system.

File: tools/spatialDraw/index.js [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+
+// eslint-disable-next-line no-undef
+main = ({width, height}) => {
+    mainData.width = width;
+    mainData.height = height;
+};
+
+let rendererStarted = false;
+
+function initRenderer() {
```
Comment (benptc) [inline tools/spatialDraw/index.js:189]: Not sure if it's worth integrating yet but a potential simplification of the boilerplate Three.js scene instantiation and rendering with the gl-proxy is to use the ThreejsInterface that @hobinjk-ptc wrote (example: https://github.com/ptcrealitylab/vuforia-spatial-analytics-addon/blob/map-tool/tools/map/index.js)

File: tools/spatialDraw/DrawingManager.js [hunk 9]
```diff
@@ -0,0 +1,758 @@
...
+    const circleResolution = 8; // 8 points
+
+    for (let i = 0; i < circleResolution; i++) {
+        circlePoints.push(new THREE.Vector2(radius * Math.sin(2 * Math.PI * i / circleResolution), radius * Math.cos(2 * Math.PI * i / circleResolution)));
+    }
+
+    return new THREE.Shape(circlePoints);
+}
+
+DrawingManager.Tool.Line = class extends DrawingManager.Tool {
```
Comment (benptc) [inline tools/spatialDraw/DrawingManager.js:416]: Nice use of inheritance, I should aim to write code as high quality as yours 😄

**Review by benptc — APPROVED** [review]
Amazing work! 🥇  Your system design and code quality continues to impress me! Left one comment about a new Three.js wrapper that you could consider using, but I'm ok if we leave that out for now since you've got everything working.

File: tools/spatialDraw/DrawingManager.js (hunk 2 above)
Comment (dangond-ptc) [reply tools/spatialDraw/DrawingManager.js:210]: Does popEvent make more sense, being the opposite of pushEvent? It does the latest event, regardless of whether or not it's a draw or erase event, so I'm not sure what you mean there.

**Pull request merged**